### DISCUSSION TITLE
Question about using the Box Provider

### DISCUSSION THREAD

**Discussion by kobay** [nextauthjs/next-auth#9777] [Help]
### Question 💬

I'm trying to use NextAuth.js v4 with Box Provider, but it's not working. 

I have tried the following.

- I cloned the example app found in the getting started guide.
- In Box, I created an app configuration, which is standard oauth2.0 user authentication, copied CLIENT_ID and CLIENT_SECRET to `.env.local` and I set `http://localhost:3000/api/auth/callback/box` for the redirect URI.
- I added the following code to `[...nextauth].ts`.

```:ts
BoxProvider({
  clientId: process.env.BOX_ID,
  clientSecret: process.env.BOX_SECRET,
  authorization: {params:{scope: undefined }}
})
```

I added `Authorization: {params: {scope: undefined }}` because NextAuth will set a query parameter `scope=openid` to authorization request and as Box doesn't understand the scope Box returns an error.

with these config, I was able to proceed until consent to grant access screen.
but after clicking the grant button, the following error occurs.

```
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error invalid_client (The client credentials are invalid) {
  error: OPError: invalid_client (The client credentials are invalid)
      at processResponse (<project home>/node_modules/.pnpm/openid-client@5.4.3/node_modules/openid-client/lib/helpers/process_response.js:38:13)
      at Client.grant (<project home>/node_modules/.pnpm/openid-client@5.4.3/node_modules/openid-client/lib/client.js:1327:22)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Client.oauthCallback (<project home>/node_modules/.pnpm/openid-client@5.4.3/node_modules/openid-client/lib/client.js:603:24)
      at async oAuthCallback (<project home>/node_modules/.pnpm/next-auth@4.22.1_next@13.4.9_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/lib/oauth/callback.js:111:16)
      at async Object.callback (<project home>/node_modules/.pnpm/next-auth@4.22.1_next@13.4.9_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/routes/callback.js:52:11)
      at async AuthHandler (<project home>/node_modules/.pnpm/next-auth@4.22.1_next@13.4.9_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/index.js:208:28)
      at async NextAuthApiHandler (<project home>/node_modules/.pnpm/next-auth@4.22.1_next@13.4.9_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/next/index.js:22:19)
      at async NextAuth._args$ (<project home>/node_modules/.pnpm/next-auth@4.22.1_next@13.4.9_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/next/index.js:106:14) {
    name: 'OAuthCallbackError',
    code: undefined
  },
  providerId: 'box',
  message: 'invalid_client (The client credentials are invalid)'
}
```

I've verified that Box returns the Auth Code, but it seems that NextAuth is failing to exchange it for the Access Token.
How can I get it to work properly?

### How to reproduce ☕️

please follow what i did 

- clone the example app found in the getting started guide.
- In Box, create an app configuration, which is standard oauth2.0 user authentication, copy CLIENT_ID and CLIENT_SECRET to `.env.local` and set `http://localhost:3000/api/auth/callback/box` for the redirect URI.
- added the following code to `[...nextauth].ts`.

```:ts
BoxProvider({
  clientId: process.env.BOX_ID,
  clientSecret: process.env.BOX_SECRET,
  authorization: {params:{scope: undefined }}
})
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by yuichiromukaiyama** [comment]
I had the exact same error about this. I investigated the cause and found the following

**my config**

Our next-auth option is as follows.

```typescript
const boxProvider: OAuthConfig<any> = BoxProvider({
  clientId: process.env.OAUTH_BOX_CLIENT_ID,
  clientSecret: process.env.OAUTH_BOX_CLIENT_SECRET,
  authorization: {
    params: {
      scope: 'root_readonly',
    },
  },
});

export const authOptions: AuthOptions = {
  providers: [boxProvider],
  pages: {
    signIn: '/auth/signin',
    signOut: '/auth/signOut',
  },
  session: {
    strategy: 'jwt',
    maxAge: 60 * 60 * 24 * 30,
  },
  secret: process.env.NEXTAUTH_SECRET,
  // Return value after OAuth contains an array and an error occurs. Comment out once.
  // adapter: prismaAdapter,
  callbacks: {},
};

export default NextAuth(authOptions);
```

1. `/oauth2/token` error

After OAuth authentication succeeds, a request is made to `/oauth2/token`. If I check the request body at this time, you will see the following.

```js
{
  grant_type: 'authorization_code',
  code: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
  redirect_uri: 'http://localhost:3000/api/auth/callback/box',
  code_verifier: undefined,
}
```

However, in the case of a box, the correct statement is as follows.

```js
{
  grant_type: 'authorization_code',
  code: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
  redirect_uri: 'http://localhost:3000/api/auth/callback/box',
  code_verifier: undefined,
  client_id: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
  client_secret: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
}
```

Does anyone know how to add a request body for requests to `/oauth2/token`?

2. Incorrect Header format for `/users/me`.

header and found the following

```js
{
  headers: {
    Authorization: 'bearer xxxxxxxxxxxxxxxxxxxxxxx,
    Accept: 'application/json'
  },
  body: undefined
}
```

When I changed the parameter to "Bearer" instead of "bearer", I was able to communicate with Box.

As described above, I was able to verify that OAuth with Box succeeded by changing two parameters. However, I did not know how to change the parameters to the Box Provider. How can I apply such parameters? (Can someone please tell me how to do this...)

**Comment by stale** [comment]
It looks like this issue did not receive any activity for 60 days. It will be closed in 7 days if no further activity occurs. If you think your issue is still relevant, commenting will keep it open. Thanks!

**Comment by calum-bird** [comment]
Did anyone involved find a way to resolve this without forking NextAuth or openid-client?

**Comment by calum-bird** [comment]
For anyone like me who was looking for closure here, I found this configuration to be functional for Box:
```typescript
BoxProvider({
      clientId: process.env.BOX_CLIENT_ID,
      clientSecret: process.env.BOX_CLIENT_SECRET,
      authorization: {
        params: { scope: undefined },
      },
      client: {
        token_endpoint_auth_method: "client_secret_post",
      },
      userinfo: {
        url: "https://api.box.com/2.0/users/me",
        async request({ tokens, provider }) {
          return await fetch("https://api.box.com/2.0/users/me", {
            headers: {
              Authorization: `Bearer ${tokens.access_token}`,
            },
          }).then(async (res) => await res.json());
        },
      },
    }),
```

The `client.token_endpoint_auth_method` instructs `openid-client`, which underpins `next-auth`, to send both the client id and client secret to the token endpoint, when set to `client_secret_post`. This solves problem 1 by @yuichiromukaiyama. As for problem 2, we use the `userinfo` parameter to send a modified request that always uses `Bearer`, instead of `bearer` which Box returns in the tokenset from problem 1.

**Reply by yuichiromukaiyama** [reply]
wonderful! I had given up on using NextAuth... 😂
